nginx/modsecurity/ngx_http_modsecurity.c

```diff
@@ -916,35 +917,27 @@ ngx_http_modsecurity_body_handler(ngx_http_request_t *r)
     ngx_http_core_run_phases(r);
 }
 
-static void
-store_azwaf_latency(ngx_http_modsecurity_ctx_t* ctx, ngx_http_variable_value_t* waf_latency_var)
-{
-    int sec, ms;
-    char* token = strtok((char*)waf_latency_var->data, ".");
-    sec = atoi(token);
-    token = strtok(NULL, " ");
-    ms = atoi(token);
-    ctx->azwaf_latency = sec * 1000 + ms;
-}
-
 static ngx_int_t
-ngx_http_calculate_modsec_latency(ngx_http_request_t *r, ngx_http_modsecurity_ctx_t* ctx)
+ngx_http_modsecurity_set_latency(ngx_http_request_t *r)
 {
-    ngx_time_t *end_time;
-    ngx_msec_int_t   ms;
-    ngx_time_update();
-    end_time = ngx_timeofday();
+    ngx_http_modsecurity_ctx_t *ctx;
+    ngx_http_variable_value_t  *modsec_latency_var;
+    ngx_msec_t                  azwaf_latency;
+
+    ctx = ngx_http_get_module_ctx(r, ngx_http_modsecurity);
+    if (NULL == ctx) {
+        return NGX_ERROR;
+    }
 
-    ms = (ngx_msec_int_t)
-        ((end_time->sec - ctx->start_time->sec) * 1000 + (end_time->msec - ctx->start_time->msec));
-    ms = ngx_max(ms, 0);
+    ctx->latency = ngx_current_msec - ctx->start_time;
 
-    // Add azwaf latency if we are in hybrid mode
-    if (ctx->azwaf_latency != 0) {
-        ms += ctx->azwaf_latency;
+    azwaf_latency = 0;
+    if (ngx_http_azwaf_get_latency(r, &azwaf_latency) == NGX_OK) {
```

Comment:
Can azwaf_latency be a negative number due to wrap around?

Reply:
I am concerned that a wrapped around value would get added to the latency. SO we won't see buffer overrun but the numbers would still look weird right?

Reply:
My previous point was that the wrap around is not going to happen for ~584,942,417 years. Not something we have to worry about.

Reply:
Oh I am not concerned about actual wrap around. I am more concerned if there is a bug in the get function and we get a negative value :) If that is not possible, then we are fine here.

Reply:
The variable and arguments are of ngx_msec_t type. This finally translates to an 8 byte unsigned value. The reason negatives were possible earlier was because the type was ngx_msec_int_t which is an 8 byte signed value. To confirm this

(gdb) p (ngx_msec_t)-1
$1 = 18446744073709551615
(gdb) p (ngx_msec_int_t)-1
$2 = -1